src/plugins/syscalls/syscalls.cpp

```diff
@@ -441,13 +510,45 @@ static event_response_t win_cb(drakvuf_t drakvuf, drakvuf_trap_info_t* info)
         }
     }
 
-    print_header(s->format, drakvuf, info);
-    if ( nargs )
+    // store informations we need to pass to win_ret_cb()
+    wrapper->args = (unsigned char*)g_malloc(sizeof(char)*size);
+    memcpy(wrapper->args, buf, sizeof(char)*size);
+    wrapper->wsc = wsc;
+    wrapper->target_cr3 = info->regs->cr3;
+    if ( !drakvuf_get_current_thread_id(drakvuf, info->vcpu, &wrapper->target_tid) )
+        goto exit;
+
+    // read RSP to get the return address and add a trap on it
+    ctx.addr = info->regs->rsp;
+    if ( VMI_FAILURE == vmi_read(vmi, &ctx, s->reg_size, &ret_addr, NULL) )
+        goto exit;
+
+    PRINT_DEBUG("Adding trap to return address 0x%lx\n", s->ret_addr);
+
+    // Most of the time the return address will be the same for a given syscall.
+    // If it is not, then we update the return address of that syscall and add a
+    // new trap.
+    if (ret_addr != s->ret_addr)
     {
-        print_nargs(s->format, nargs);
-        print_args(s, drakvuf, info, wsc, buf);
+        s->ret_addr = ret_addr;
+        drakvuf_trap_t* trap = (drakvuf_trap_t*)g_malloc0(sizeof(drakvuf_trap_t));
```

Comment:
Where is trap getting freed? Valgrind reports memory leak here

Reply:
I find that if add a return address trap in this time,when one syscall occurs, multiple callbacks are triggered and 
it will accumulated? Why is that? if that add this code "drakvuf_remove_trap(drakvuf, info->trap, NULL);" in callbacks?

Reply:
> Where is trap getting freed? Valgrind reports memory leak here

I find that if add a return address trap in this time,when one syscall occurs, multiple callbacks are triggered and 
it will accumulated? Why is that? if that add this code "drakvuf_remove_trap(drakvuf, info->trap, NULL);" in callbacks?